subsys/net/lib/nrf_cloud/src/nrf_cloud_logs.c

```diff
@@ -0,0 +1,452 @@
+	 * a JSON array. Close the array then send it.
+	 */
+	if ((num_msgs != 0) && (log_format_current == LOG_OUTPUT_TEXT)) {
+		ring_buf_put(&log_nrf_cloud_rb, "]", 1);
+	}
+
+	stored = ring_buf_size_get(&log_nrf_cloud_rb);
+	output.data.len = ring_buf_get_claim(&log_nrf_cloud_rb,
+					     (uint8_t **)&output.data.ptr,
+					     stored);
+	if (output.data.len != stored) {
+		LOG_WRN("Capacity:%u, free:%u, stored:%u, claimed:%u",
+			ring_buf_capacity_get(&log_nrf_cloud_rb),
+			ring_buf_space_get(&log_nrf_cloud_rb),
+			stored, output.data.len);
+		stored = output.data.len;
+	}
+
+	if ((log_format_current == LOG_OUTPUT_DICT) &&
+	    IS_ENABLED(CONFIG_NRF_CLOUD_LOG_DICT_JSON)) {
+		/* Convert to base64 and wrap in JSON */
+		size_t olen;
+		size_t buflen;
+
+		/* Determine buffer size needed */
+		base64_encode(NULL, 0, &olen, output.data.ptr, output.data.len);
+		buflen = strlen(JSON_FMT1) + olen + strlen(JSON_FMT2);
+
+		base64_buf = k_calloc(buflen, 1);
+		if (base64_buf == NULL) {
+			err = -ENOMEM;
+			goto cleanup;
+		}
+
+		memcpy(base64_buf, JSON_FMT1, strlen(JSON_FMT1));
+		err = base64_encode(&base64_buf[strlen(JSON_FMT1)], olen, &olen,
+				    output.data.ptr, output.data.len);
+		if (err) {
+			goto cleanup;
+		}
+		memcpy(&base64_buf[strlen(JSON_FMT1) + olen], JSON_FMT2, strlen(JSON_FMT2));
+		output.data.ptr = base64_buf;
+		output.data.len = buflen - 1;
+	} /* Else send ring buffer contents as is -- JSON or raw binary. */
+
+	if (IS_ENABLED(CONFIG_NRF_CLOUD_MQTT)) {
+		err = nrf_cloud_send(&output);
+	} else if (IS_ENABLED(CONFIG_NRF_CLOUD_REST)) {
+		/* send to d2c topic, unless bulk is true, in which case d2c/bulk is used */
+		err = nrf_cloud_rest_send_device_message(context.rest_ctx,
+							 context.device_id,
+							 output.data.ptr, false, NULL);
+	} else {
+		err = -ENODEV;
+	}
+	if (!err) {
+		stats.lines_sent += num_msgs;
+		stats.bytes_sent += output.data.len;
+	}
+
+cleanup:
+	if (base64_buf) {
+		k_free(base64_buf);
+	}
+	ret = ring_buf_get_finish(&log_nrf_cloud_rb, stored);
+	ring_buf_reset(&log_nrf_cloud_rb);
+	num_msgs = 0;
+
+	if (ret) {
+		LOG_ERR("Error finishing ring buffer: %d", ret);
+		err = ret;
+	}
+	return err;
+}
+
+static int logger_out(uint8_t *buf, size_t size, void *ctx)
+{
+	ARG_UNUSED(ctx);
+	int err = 0;
+	struct nrf_cloud_data data;
+	size_t orig_size = size;
+	uint32_t stored;
+	int extra;
+
+	if (!size) {
+		return 0;
+	}
+
+	if (log_format_current == LOG_OUTPUT_TEXT) {
+		extra = 2;
+		if (context.src_name) {
+			int len = strlen(context.src_name);
+
+			/* We want to pass the log source as a separate field for cloud filtering.
+			 * Because the Zephyr logging subsystem prefixes the log text with
+			 * 'src_name: ', remove it before encoding the log.
+			 */
+			if ((strncmp(context.src_name, (const char *)buf, MIN(size, len)) == 0) &&
+			    ((len + 2) < size) && (buf[len] == ':') && (buf[len + 1] == ' ')) {
+				buf += len + 2;
+				size -= len + 2;
+			}
```

Comment:
Another approach might be to move all the JSON formatting logic out of `logger_out` and into `logger_process`, and let `logger_out` focus solely on sending device message strings.

This aligns better with the "formatter" vs. "printer" separation that Zephyr appears to be trying to establish between the output and process funcs, but it has the downside that for raw-binary printing, you'll be sending binary strings to an API that technically is supposed to expect ASCII. So I'm kinda on the fence about this idea, but it seems worth mentioning

Reply:
I can't do the first suggestion because the Zephyr log_output.c function which generates the prefix on each log line does that when `logger_process()` calls `log_output_func()`.

Same for your idea of moving JSON formatting into `logger_process()`.

Reply:
I was thinking that instead of calling `log_output_func`, you could call zephyr's `log_output_process` directly (this function is what `log_output_func` winds up calling). This would allow you to pass custom flags and customize the behavior

But you'd have to check for dict mode, and in that scenario, call `log_dict_output_msg_process` instead. So maybe the fragility of dipping into semi-internal Zephyr functions (they have exposed header defs at least) is not worth avoiding parsing the data out of the printed message. Up to you

If you did this, then it would be possible to move the JSON formatting into logger_process, since you would have access to the raw message (without prefix) as the string passed into `logger_out`, and you could acquire the module name, etc, by inspecting `msg->log`

Thinking longer on it, I wonder if it would make sense to, rather than use the existing log format functions, obtained via `log_format_func_t_get`, instead write our own format function that operates directly on the `log_output->log` struct like the built-in ones do.

This is justifiable since what we are doing is essentially inventing a JSON output format

Reply:
No